### PR TITLE
Rust BPF SDK

### LINKED ISSUE #4981: Rust BPF: Add Syscall utilities
#### Problem

Users have to read and parse the syscall accounts themselves.

#### Proposed Solution

Provide utilties that will return the data in an easy to use format.

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

Rust BPF programs rely on types and helpers based on the Rust `no_std` requirements

#### Summary of Changes

These changes are a work in progress and need review/feedback

What this PR does:
- Renames, cleanups and consolidates the Rust BPF SDK crates
- Pulls in pieces of the Solana SDK so that common types are used

Problems this PR has:
- Not everything in the Solana SDK is compatible with, relevant to, or needed by Rust BPF programs.  So only a subset of the Solana SDK is made available to Rust BPF programs and that includes modifications to some of the files.
- The subset of files is currently duplicated in the tree, probably will require splitting up the Solana SDK.
- KeyedAccounts are a very unfriendly structure to deserialize and pass to programs.  Instead, these changes define an `Account` type that is more friendly from both a usability and performance standpoint.  This is great for BPF programs, bad for duplication or confusion between the two SDKs
- `sysvar/fees.rs` requires `FeeCalculator` but `FeeCalculator` uses signed division which BPF has no support for.
- Looks like `sysvar/stake_history.rs` uses `HashMap` which I don't think is deterministic, may need to use a different data structure

Fixes #4981 #3111

### CODE REVIEW

**Comment by jackcmay** [comment]
@mvines, @rwalker I'd love to get your eyes on these changes.  This is a first swipe, so looking for feedback and ideas to address the problems listed in the description.

File: programs/bpf/rust/clock/Cargo.toml [hunk 1]
```diff
@@ -2,22 +2,21 @@
 # Note: This crate must be built using build.sh
 
 [package]
-name = "solana-bpf-rust-tick-height"
+name = "solana-bpf-rust-clock"
```
Comment (mvines) [inline programs/bpf/rust/clock/Cargo.toml]: Can I tempt you with `programs/bpf/rust/128bit/Cargo.toml` -> `programs/bpf-rust-128bit/Cargo.toml`

Comment (jackcmay) [reply programs/bpf/rust/clock/Cargo.toml]: No sweat, also means `programs/bpf-c/...`.  Mind if I do that in a different PR?

**Comment by mvines** [comment]
Before this PR gets bigger, can we make it smaller?  Mixing renaming and functional changes is pretty hard on reviewers

**Comment by rob-solana** [comment]
the duplication looks like super-bad juju.  if bpf is a "real" SDK, the non-BPF SDK can pull from the "real" one

sdk::timing needs to die or be split

**Comment by jackcmay** [comment]
The changes in this PR pretty much go hand-in-hand and for the most part should be pretty straight forward to review.  I'd rather not incur the overhead of reworking these changes until I get feedback and a better direction as to what we want to do about the problems mentioned in the description.

**Comment by jackcmay** [comment]
@rob-solana Yup, the duplication is not what we will end up with, but before Ii spend a lot of time splitting the two I want feedback on how we want to split it up.

**Comment by mvines** [comment]
Can `sdk/src/` support all Rust targets?   We don't have `sdk/macos/rust/src` or `sdk/linux/rust/src` either.   The duplication in `sdk/bpf/rust` looks nasty

**Comment by mvines** [comment]
Perhaps a pattern like what libloading does? https://docs.rs/libloading/0.5.2/libloading/os/index.html
BPF specific stuff goes in a `solana_sdk::bpf` module of the single sdk crate?

**Comment by jackcmay** [comment]
The `Account` definition, signed division, time, rand, and possibly some dependent crates are the things I'm aware of that might not be compatible in the existing full SDK.  But, it also becomes less clear to developers which pieces are relevant to on-chain programs.  Things like message, packet, rpc_port, transport, .... are noise.

**Comment by rob-solana** [comment]
fee_calculator can be reworked to not use signed division

**Pull request closed without merging**